ring/partition_instance_lifecycler_test.go

```diff
@@ -0,0 +1,336 @@
+		lifecycler2a := NewPartitionInstanceLifecycler(lifecycler2aConfig, "test", ringKey, store, logger, nil)
+		require.NoError(t, services.StartAndAwaitRunning(ctx, lifecycler1a))
+		require.NoError(t, services.StartAndAwaitRunning(ctx, lifecycler1b))
+		require.NoError(t, services.StartAndAwaitRunning(ctx, lifecycler2a))
+		t.Cleanup(func() {
+			// Ensure we stop all lifecyclers once the test is terminated.
+			require.NoError(t, services.StopAndAwaitTerminated(ctx, lifecycler1a))
+			require.NoError(t, services.StopAndAwaitTerminated(ctx, lifecycler1b))
+			require.NoError(t, services.StopAndAwaitTerminated(ctx, lifecycler2a))
+		})
+
+		actual := getPartitionRingFromStore(t, store, ringKey)
+		assert.Len(t, actual.Partitions, 2)
+		assert.True(t, actual.HasPartition(1))
+		assert.True(t, actual.HasPartition(2))
+		assert.Equal(t, map[int32][]string{1: {"instance-zone-a-1", "instance-zone-b-1"}, 2: {"instance-zone-a-2"}}, actual.ownersByPartition())
+
+		// Switch partition 1 to inactive.
+		require.NoError(t, lifecycler1a.ChangePartitionState(ctx, PartitionInactive))
+
+		// Wait longer than deletion wait period. We expect that the partition is not
+		// delete yet because there are still owners.
+		time.Sleep(2 * lifecycler1aConfig.DeleteInactivePartitionAfterDuration)
+
+		actual = getPartitionRingFromStore(t, store, ringKey)
+		assert.Len(t, actual.Partitions, 2)
+		assert.True(t, actual.HasPartition(1))
+		assert.True(t, actual.HasPartition(2))
+		assert.Equal(t, map[int32][]string{1: {"instance-zone-a-1", "instance-zone-b-1"}, 2: {"instance-zone-a-2"}}, actual.ownersByPartition())
+
+		// Stop instance-zone-a-1.
+		lifecycler1a.SetRemoveOwnerOnShutdown(true)
+		require.NoError(t, services.StopAndAwaitTerminated(ctx, lifecycler1a))
+
+		actual = getPartitionRingFromStore(t, store, ringKey)
+		assert.Len(t, actual.Partitions, 2)
+		assert.True(t, actual.HasPartition(1))
+		assert.True(t, actual.HasPartition(2))
+		assert.Equal(t, map[int32][]string{1: {"instance-zone-b-1"}, 2: {"instance-zone-a-2"}}, actual.ownersByPartition())
+
+		// Stop instance-zone-b-1.
+		lifecycler1b.SetRemoveOwnerOnShutdown(true)
+		require.NoError(t, services.StopAndAwaitTerminated(ctx, lifecycler1b))
+
+		actual = getPartitionRingFromStore(t, store, ringKey)
+		assert.Equal(t, map[int32][]string{2: {"instance-zone-a-2"}}, actual.ownersByPartition())
+
+		// We expect remaining lifecyclers to clean up the inactive partition now.
+		assert.Eventually(t, func() bool {
+			actual := getPartitionRingFromStore(t, store, ringKey)
+			return !actual.HasPartition(1)
+		}, time.Second, eventuallyTick)
+
+		// Partition 2 should still exist.
+		actual = getPartitionRingFromStore(t, store, ringKey)
+		assert.True(t, actual.HasPartition(2))
+	})
+
+	t.Run("should not create the partition but wait until exist in the ring if lifecycler has been configured to not create partition at startup", func(t *testing.T) {
```

Comment:
```suggestion
	t.Run("should not create the partition but wait until partition exists in the ring if lifecycler has been configured to not create partition at startup", func(t *testing.T) {
```

Reply:
Done in https://github.com/grafana/dskit/pull/483/commits/1069f18ad7601d8eeaaef953f7cbed4d4e481777